src/jupyterhub_config/custom_kube_spawner.py

```diff
@@ -0,0 +1,345 @@
+
+        bashrc_dest = user_dir / '.bashrc'
+        bash_profile_dest = user_dir / '.bash_profile'
+
+        if not bashrc_dest.exists():
+            self.log.info(f'Creating .bashrc file for {user_dir}')
+            shutil.copy2(bashrc_tmpl, bashrc_dest)
+
+        if not bash_profile_dest.exists():
+            self.log.info(f'Creating .bash_profile file for {user_dir}')
+            shutil.copy2(bash_profile_tmpl, bash_profile_dest)
+
+    def _ensure_user_jupyter_directory(self, user_dir: Path):
+        """
+        Create the user's Jupyter directory and subdirectories if they do not exist. And set the
+        environment variables for Jupyter to use these directories.
+        """
+
+        if not user_dir.exists():
+            raise ValueError(f'User directory {user_dir} does not exist')
+
+        jupyter_dir = user_dir / '.jupyter'
+        jupyter_runtime_dir = jupyter_dir / 'runtime'
+        juputer_data_dir = jupyter_dir / 'data'
+
+        jupyter_dir.mkdir(parents=True, exist_ok=True)
+        jupyter_runtime_dir.mkdir(parents=True, exist_ok=True)
+        juputer_data_dir.mkdir(parents=True, exist_ok=True)
+
+        # copy the jupyter_jupyter_ai_config.json file to the user's .jupyter directory
+        # ref: https://jupyter-ai.readthedocs.io/en/latest/users/index.html#configuring-as-a-config-file
+        jupyter_notebook_config = Path(os.environ['JUPYTERHUB_CONFIG_DIR']) / os.environ['JUPYTER_AI_CONFIG_FILE']
+        shutil.copy2(jupyter_notebook_config, jupyter_dir / os.environ['JUPYTER_AI_CONFIG_FILE'])
+
+        self.environment['JUPYTER_CONFIG_DIR'] = str(jupyter_dir)
+        self.environment['JUPYTER_RUNTIME_DIR'] = str(jupyter_runtime_dir)
+        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
+
+    def _ensure_virtual_environment(self, user_env_dir: Path):
+        """
+        Ensure the user's virtual environment exists. If it does not exist, it is
+        created with the system site-packages included.
+        """
+        if not user_env_dir.exists():
+            user_env_dir.mkdir(parents=True, exist_ok=True)
+            self.log.info(f'Creating virtual environment for {self.user.name}')
+            try:
+                # Create a virtual environment with system site-packages access
+                venv.create(env_dir=user_env_dir, system_site_packages=True, with_pip=True)
+            except Exception as e:
+                raise ValueError(f'Failed to create virtual environment for {self.user.name}: {e}') from e
+        else:
+            self.log.info(f'Reusing virtual environment for {self.user.name}')
+
+    def _configure_environment(self, user_dir: Path, user_env_dir: Path, username: str):
+        self.environment.update({key: str(value) for key, value in os.environ.items() if key not in self.environment})
+        self.environment['JUPYTER_MODE'] = 'jupyterhub-singleuser'
+        # self.environment['JUPYTERHUB_ADMIN'] = str(self.user.admin)
```

Comment:
I don't totally understand why this is not longer needed

Reply:
I put it back. `self.user.admin` is a bool and gave me trouble before. I commented it and forget to comment it back. It's current just an indicator. Admin user is set by the custom auth now via a KBase Auth role.

Reply:
So it is necessary, or can it be removed completely?

Reply:
we can keep it for troubleshooting I guess. It's not strictly needed.

Reply:
If it's not necessary I'd probably remove it but I leave it up to you

Reply:
I will leave it in case we need it. KubeSpawner is pretty fragile now.

Reply:
But if you keep it, did you mean to stringify the bool?

Reply:
Yes. non-string value is not allowed to save to `self.environment` for kubespawner.

Reply:
is it used anywhere else where it's assumed to be a bool? E.g. now `if self.environment['JUPYTERHUB_ADMIN']` will always be True

Reply:
Not in our codebase. Inside of spawner, I think it might just use `self.user.admin`.

Reply:
:+1: